Maple2.Server.Game/Manager/ShopManager.cs

```diff
@@ -85,162 +85,211 @@ private bool TryGetShopData(int shopId, [NotNullWhen(true)] out CharacterShopDat
+            long restockTime = GetRestockTime(metadata);
+            return CreateInstancedShop(metadata, CreateShopData(metadata, restockTime), restockTime);
+        }
 
-        if (!TryGetShopData(serverShop.Id, out CharacterShopData? data)) {
-            long restockTime = GetRestockTime(serverShop.RestockData!.Interval);
-            if (serverShop.RestockTime > 0) {
-                restockTime = serverShop.RestockTime;
+        if (data.RestockTime < DateTime.Now.ToEpochSeconds()) {
+            using GameStorage.Request db = session.GameStorage.Context();
+            if (!TryRemoveShopData(data.ShopId)) {
+                throw new InvalidOperationException($"Failed to remove shop data {metadata.Id} for character {session.CharacterId}");
             }
-            return CreateInstancedShop(serverShop, CreateShopData(serverShop), restockTime);
+            db.DeleteCharacterShopData(metadata.RestockData.AccountWide ? session.AccountId : session.CharacterId, metadata.Id);
 
-        }
-        if (data.RestockTime < DateTime.Now.ToEpochSeconds()) {
-            return CreateInstancedShop(serverShop, CreateShopData(serverShop), GetRestockTime(serverShop.RestockData!.Interval));
+            long restockTime = GetRestockTime(metadata);
+            return CreateInstancedShop(metadata, CreateShopData(metadata, restockTime), restockTime);
         }
 
         // Assemble shop
-        instancedShop = serverShop.Clone();
-        instancedShop.RestockTime = data.RestockTime;
+        instancedShop = new Shop(metadata) {
+            RestockTime = data.RestockTime,
+            Items = GetShopItems(metadata),
+        };
 
         if (!TryGetShopItemData(instancedShop.Id, out IDictionary<int, CharacterShopItemData>? dataDictionary)) {
-            dataDictionary = CreateShopItemData(serverShop);
+            dataDictionary = CreateShopItemData(instancedShop);
         }
 
         foreach ((int shopItemId, CharacterShopItemData itemData) in dataDictionary) {
-            if (!serverShop.Items.TryGetValue(shopItemId, out ShopItem? shopItem)) {
+            if (!instancedShop.Items.TryGetValue(shopItemId, out ShopItem? shopItem)) {
                 continue;
             }
 
-            ShopItem itemCopy = shopItem.Clone();
-            itemCopy.Item = itemData.Item;
-            itemCopy.StockPurchased = itemData.StockPurchased;
-            instancedShop.Items.Add(shopItemId, itemCopy);
+            instancedShop.Items[shopItemId] = new ShopItem(shopItem.Metadata) {
+                Item = itemData.Item,
+                StockPurchased = itemData.StockPurchased,
+            };
         }
 
         instancedShops[instancedShop.Id] = instancedShop;
         return instancedShop;
     }
 
-    private Shop CreateInstancedShop(Shop shop, CharacterShopData shopData, long restockTime) {
-        Shop instancedShop = shop.Clone()!;
-
+    private Shop CreateInstancedShop(ShopMetadata metadata, CharacterShopData shopData, long restockTime) {
```

Comment:
**suggestion:** New dynamic approach to shop item generation

This new approach allows for more varied shop contents. Consider adding comments to explain the logic behind item generation and any measures taken to ensure fairness and consistency in the randomness.